libgnucash/engine/gnc-commodity.cpp

```diff
@@ -351,24 +347,13 @@ gnc_quote_source_init_tables (void)
 gnc_quote_source *
 gnc_quote_source_add_new (const char *source_name, gboolean supported)
 {
-    gnc_quote_source *new_source;
-
-    DEBUG("Creating new source %s", (source_name == NULL ? "(null)" : source_name));
-    new_source = malloc(sizeof(gnc_quote_source));
-    new_source->supported = supported;
-    new_source->type = SOURCE_UNKNOWN;
-    new_source->index = g_list_length(new_quote_sources);
-
+    DEBUG("Creating new source %s", (!source_name ? "(null)" : source_name));
     /* This name can be changed if/when support for this price source is
      * integrated into gnucash. */
-    new_source->user_name = g_strdup(source_name);
-
     /* This name is permanent and must be kept the same if/when support
      * for this price source is integrated into gnucash (i.e. for a
      * nice user name). */
-    new_source->internal_name = g_strdup(source_name);
-    new_quote_sources = g_list_append(new_quote_sources, new_source);
-    return new_source;
+    return &new_quote_sources.emplace_back (supported, SOURCE_UNKNOWN, source_name, source_name);
```

Comment:
You're returning a reference to an object that can be moved when a vector reallocation occurs.

* https://en.cppreference.com/w/cpp/container/vector/push_back
* https://bugs.gnucash.org/show_bug.cgi?id=799305

Reply:
I'm struggling a bit to find an elegant solution for this. I don't think it's nice to do vector.reserve(100000) 😭

Reply:
100000 would be silly. 100 would be sufficient, F::Q only has 54, but it's still an ugly and brittle way to do it.

Reply:
@nomis does the bug disappear when you redefine QuoteSourceVec as std::list<quote_source_info>?

Reply:
@nomis #1936

Reply:
> @nomis does the bug disappear when you redefine QuoteSourceVec as std::list<quote_source_info>?

It would do, because std::list doesn't invalidate references when the list changes.